pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -259,8 +267,16 @@ func NewKV(cfg KVConfig) (*KV, error) {
 		casBroadcastsEnabled: atomic.NewBool(true),
 	}
 
+	mlkv.createAndRegisterMetrics()
+
+	for _, c := range cfg.Codecs {
+		mlkv.codecs[c.CodecID()] = c
+	}
+
 	mlCfg.Delegate = mlkv
 
+	// Wait for memberlist and broadcasts
+	mlkv.initWG.Add(1)
```

Comment:
What is this `unitWG` used for? The `Done()` is called below before returning so I can't see how any function on this receiver may actually hit the case there's something to wait.

Reply:
This actually solves real race... as soon as `memberlist.Create` is called, memberlist may start calling [delegate methods](https://github.com/cortexproject/cortex/blob/33944e9bfa25855605b62f0083129285a1b2736b/vendor/github.com/hashicorp/memberlist/delegate.go) if it receives traffic (those that call `initWG.Wait()`), but delegate initialization (`mlkv` here) is not finished yet.

Reply:
Thanks for the explaination. What do you think @periklis to enhance the comment to explain it? What Peter said is quite clear to me (which I'm not much expert of this gossip integration).